File: syntax/tests/idempotency/bucklescript/belt/belt_Set.ml
(* Copyright (C) 2017 Authors of BuckleScript
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * In addition to the permissions granted to you by the LGPL, you may combine
 * or link a "work that uses the Library" with a publicly distributed version
 * of this file to produce a combined library or application, then distribute
 * that combined work under the terms of your choosing, with no requirement
 * to comply with the obligations normally placed on you by section 4 of the
 * LGPL version 3 (or the corresponding section of a later version of the LGPL
 * should you choose to use a later version).
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. *)

module Int = Belt_SetInt
module String = Belt_SetString
module Dict = Belt_SetDict
  
module A = Belt_Array


type ('value, 'id) id = ('value, 'id) Belt_Id.comparable
type ('value, 'id ) cmp = ('value, 'id) Belt_Id.cmp


type ('value, 'id) t = {
    cmp: ('value, 'id) cmp;
    data: ('value, 'id) Dict.t;
  } [@@bs.deriving abstract]



    
let fromArray (type value) (type identity) data ~(id : (value,identity) id)  = 
  let module M = (val id ) in
  let cmp = M.cmp in 
  t ~cmp ~data:(Dict.fromArray ~cmp data)

let remove m e =      
  let cmp, data  = cmpGet m, dataGet m  in
  let newData = Dict.remove ~cmp data e in 
  if newData == data then m 
  else t ~cmp ~data:newData  

let add m e =   
  let cmp, data = cmpGet m, dataGet m in
  let newData = Dict.add ~cmp  data e in 
  if newData == data then m 
  else 
    t ~cmp ~data:newData

let mergeMany m e =
  let cmp = cmpGet m in 
  t ~cmp  ~data:(Dict.mergeMany ~cmp (dataGet m) e )

let removeMany  m e = 
  let cmp = cmpGet m in 
  t ~cmp ~data:(Dict.removeMany ~cmp (dataGet m) e)

let union m n =   
  let cmp = cmpGet m in 
  t ~data:(Dict.union ~cmp (dataGet m) (dataGet n)) ~cmp

let intersect m n =
  let cmp = cmpGet m in 
  t ~data:(Dict.intersect ~cmp (dataGet m) (dataGet n)) ~cmp

let diff m n =
  let cmp = cmpGet m in 
  t ~cmp ~data:(Dict.diff ~cmp (dataGet m) (dataGet n))

let subset m n =     
  let cmp = cmpGet m in 
  Dict.subset ~cmp (dataGet m) (dataGet n) 

let split m e = 
  let cmp = cmpGet m in  
  let (l,  r), b = Dict.split ~cmp (dataGet m) e in 
  (t ~cmp ~data:l, t ~cmp ~data:r), b
  
let make (type value) (type identity) ~(id : (value, identity) id) =
  let module M = (val id) in 
  t ~cmp:M.cmp  ~data:Dict.empty

let isEmpty m = Dict.isEmpty (dataGet m)

let cmp m n =
  let cmp = cmpGet m in 
  Dict.cmp ~cmp (dataGet m) (dataGet n)

let eq m n =     
  Dict.eq ~cmp:(cmpGet m) (dataGet m) (dataGet n)    

let forEachU m f  = Dict.forEachU (dataGet m) f 
let forEach m  f = forEachU m (fun [@bs] a -> f a)

let reduceU m acc f = Dict.reduceU (dataGet m) acc f
let reduce m acc f = reduceU m acc (fun [@bs] a b -> f a b)
    
let everyU m f  = Dict.everyU  (dataGet m) f
let every m f  = everyU m (fun [@bs] a -> f a)
    
let someU m f = Dict.someU  (dataGet m) f
let some m f = someU m (fun [@bs] a -> f a)

let keepU m f  = 
  t ~cmp:(cmpGet m) ~data:(Dict.keepU (dataGet m) f )
let keep m f = keepU m (fun [@bs] a -> f a)
    
let partitionU m f  = 
  let l,r = Dict.partitionU (dataGet m) f in
  let cmp = cmpGet m in 
  t ~data:l ~cmp, t ~data:r ~cmp
let partition m f = partitionU m (fun [@bs] a -> f a)
    
let size m = Dict.size (dataGet m) 
let toList m = Dict.toList (dataGet m)
let toArray m = Dict.toArray (dataGet m)

let minimum m = Dict.minimum (dataGet m)
let minUndefined m = Dict.minUndefined (dataGet m) 
let maximum m = Dict.maximum (dataGet m)
let maxUndefined m = Dict.maxUndefined (dataGet m)


let get m e =   
  Dict.get ~cmp:(cmpGet m) (dataGet m) e

let getUndefined m e =   
  Dict.getUndefined ~cmp:(cmpGet m) (dataGet m) e

let getExn m e =   
  Dict.getExn ~cmp:(cmpGet m) (dataGet m) e

let has m e = 
  Dict.has ~cmp:(cmpGet m) (dataGet m) e

let fromSortedArrayUnsafe (type value) (type identity) xs ~(id : (value,identity) id ) =
  let module M = (val id) in 
  t ~cmp:M.cmp ~data:(Dict.fromSortedArrayUnsafe xs)

let getData = dataGet

let getId (type value) (type identity) (m : (value,identity) t) : (value, identity) id =
  let module T = struct
    type nonrec identity = identity
    type nonrec t = value
    let cmp =  cmpGet m
  end in
  (module T)
  
let packIdData (type value) (type identity) ~(id : (value, identity) id) ~data  =
  let module M = (val id) in 
  t ~cmp:M.cmp ~data

let checkInvariantInternal d = Dict.checkInvariantInternal (dataGet d)

File: lib/js/belt_Set.js
'use strict';

var Curry = require("./curry.js");
var Belt_SetDict = require("./belt_SetDict.js");

function fromArray(data, id) {
  var cmp = id.cmp;
  return {
          cmp: cmp,
          data: Belt_SetDict.fromArray(data, cmp)
        };
}

function remove(m, e) {
  var data = m.data;
  var cmp = m.cmp;
  var newData = Belt_SetDict.remove(data, e, cmp);
  if (newData === data) {
    return m;
  } else {
    return {
            cmp: cmp,
            data: newData
          };
  }
}

function add(m, e) {
  var data = m.data;
  var cmp = m.cmp;
  var newData = Belt_SetDict.add(data, e, cmp);
  if (newData === data) {
    return m;
  } else {
    return {
            cmp: cmp,
            data: newData
          };
  }
}

function mergeMany(m, e) {
  var cmp = m.cmp;
  return {
          cmp: cmp,
          data: Belt_SetDict.mergeMany(m.data, e, cmp)
        };
}

function removeMany(m, e) {
  var cmp = m.cmp;
  return {
          cmp: cmp,
          data: Belt_SetDict.removeMany(m.data, e, cmp)
        };
}

function union(m, n) {
  var cmp = m.cmp;
  return {
          cmp: cmp,
          data: Belt_SetDict.union(m.data, n.data, cmp)
        };
}

function intersect(m, n) {
  var cmp = m.cmp;
  return {
          cmp: cmp,
          data: Belt_SetDict.intersect(m.data, n.data, cmp)
        };
}

function diff(m, n) {
  var cmp = m.cmp;
  return {
          cmp: cmp,
          data: Belt_SetDict.diff(m.data, n.data, cmp)
        };
}

function subset(m, n) {
  var cmp = m.cmp;
  return Belt_SetDict.subset(m.data, n.data, cmp);
}

function split(m, e) {
  var cmp = m.cmp;
  var match = Belt_SetDict.split(m.data, e, cmp);
  var match$1 = match[0];
  return [
          [
            {
              cmp: cmp,
              data: match$1[0]
            },
            {
              cmp: cmp,
              data: match$1[1]
            }
          ],
          match[1]
        ];
}

function make(id) {
  return {
          cmp: id.cmp,
          data: undefined
        };
}

function isEmpty(m) {
  return Belt_SetDict.isEmpty(m.data);
}

function cmp(m, n) {
  var cmp$1 = m.cmp;
  return Belt_SetDict.cmp(m.data, n.data, cmp$1);
}

function eq(m, n) {
  return Belt_SetDict.eq(m.data, n.data, m.cmp);
}

function forEachU(m, f) {
  Belt_SetDict.forEachU(m.data, f);
}

function forEach(m, f) {
  Belt_SetDict.forEachU(m.data, Curry.__1(f));
}

function reduceU(m, acc, f) {
  return Belt_SetDict.reduceU(m.data, acc, f);
}

function reduce(m, acc, f) {
  return reduceU(m, acc, Curry.__2(f));
}

function everyU(m, f) {
  return Belt_SetDict.everyU(m.data, f);
}

function every(m, f) {
  return Belt_SetDict.everyU(m.data, Curry.__1(f));
}

function someU(m, f) {
  return Belt_SetDict.someU(m.data, f);
}

function some(m, f) {
  return Belt_SetDict.someU(m.data, Curry.__1(f));
}

function keepU(m, f) {
  return {
          cmp: m.cmp,
          data: Belt_SetDict.keepU(m.data, f)
        };
}

function keep(m, f) {
  return keepU(m, Curry.__1(f));
}

function partitionU(m, f) {
  var match = Belt_SetDict.partitionU(m.data, f);
  var cmp = m.cmp;
  return [
          {
            cmp: cmp,
            data: match[0]
          },
          {
            cmp: cmp,
            data: match[1]
          }
        ];
}

function partition(m, f) {
  return partitionU(m, Curry.__1(f));
}

function size(m) {
  return Belt_SetDict.size(m.data);
}

function toList(m) {
  return Belt_SetDict.toList(m.data);
}

function toArray(m) {
  return Belt_SetDict.toArray(m.data);
}

function minimum(m) {
  return Belt_SetDict.minimum(m.data);
}

function minUndefined(m) {
  return Belt_SetDict.minUndefined(m.data);
}

function maximum(m) {
  return Belt_SetDict.maximum(m.data);
}

function maxUndefined(m) {
  return Belt_SetDict.maxUndefined(m.data);
}

function get(m, e) {
  return Belt_SetDict.get(m.data, e, m.cmp);
}

function getUndefined(m, e) {
  return Belt_SetDict.getUndefined(m.data, e, m.cmp);
}

function getExn(m, e) {
  return Belt_SetDict.getExn(m.data, e, m.cmp);
}

function has(m, e) {
  return Belt_SetDict.has(m.data, e, m.cmp);
}

function fromSortedArrayUnsafe(xs, id) {
  return {
          cmp: id.cmp,
          data: Belt_SetDict.fromSortedArrayUnsafe(xs)
        };
}

function getData(m) {
  return m.data;
}

function getId(m) {
  var cmp = m.cmp;
  return {
          cmp: cmp
        };
}

function packIdData(id, data) {
  return {
          cmp: id.cmp,
          data: data
        };
}

function checkInvariantInternal(d) {
  Belt_SetDict.checkInvariantInternal(d.data);
}

var Int;

var $$String;

var Dict;

exports.Int = Int;
exports.$$String = $$String;
exports.Dict = Dict;
exports.make = make;
exports.fromArray = fromArray;
exports.fromSortedArrayUnsafe = fromSortedArrayUnsafe;
exports.isEmpty = isEmpty;
exports.has = has;
exports.add = add;
exports.mergeMany = mergeMany;
exports.remove = remove;
exports.removeMany = removeMany;
exports.union = union;
exports.intersect = intersect;
exports.diff = diff;
exports.subset = subset;
exports.cmp = cmp;
exports.eq = eq;
exports.forEachU = forEachU;
exports.forEach = forEach;
exports.reduceU = reduceU;
exports.reduce = reduce;
exports.everyU = everyU;
exports.every = every;
exports.someU = someU;
exports.some = some;
exports.keepU = keepU;
exports.keep = keep;
exports.partitionU = partitionU;
exports.partition = partition;
exports.size = size;
exports.toArray = toArray;
exports.toList = toList;
exports.minimum = minimum;
exports.minUndefined = minUndefined;
exports.maximum = maximum;
exports.maxUndefined = maxUndefined;
exports.get = get;
exports.getUndefined = getUndefined;
exports.getExn = getExn;
exports.split = split;
exports.checkInvariantInternal = checkInvariantInternal;
exports.getData = getData;
exports.getId = getId;
exports.packIdData = packIdData;
/* No side effect */
